File: 21-30/23. Merge k Sorted Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
    let mergeLists = (a, b) => {
        const ans = new ListNode();
        let temp = ans;
        while (a != null && b != null) {
            if (a.val < b.val) {
                temp.next = a;
                a = a.next;
            } else {
                temp.next = b;
                b = b.next;
            }
            temp = temp.next;
        }
        if (a == null) {
            temp.next = b;
        }
        if (b == null) {
            temp.next = a;
        }
        return ans.next;
    };

    if (lists.length == 0) {
        return null;
    }
    while (lists.length > 1) {
        let a = lists.shift();
        let b = lists.shift();
        let temp = mergeLists(a, b);
        lists.push(temp);
    }
    return lists[0];
};

File: 21-30/23. Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 1:
            return lists[0]
        if len(lists) == 0:
            return None
        
        while len(lists) >= 2:
            first = lists.pop()
            second = lists.pop()

            dummy = ListNode()
            cur = dummy
            while first and second:
                if first.val < second.val:
                    cur.next = ListNode(first.val)
                    first = first.next
                else:
                    cur.next = ListNode(second.val)
                    second = second.next
                cur = cur.next
            
            if first:
                cur.next = first
            else:
                cur.next = second
            
            lists.append(dummy.next)
        
        return lists[0]

